### PR TITLE
distribute oauth-server trust via a openshift-config-managed configmap

### PR DESCRIPTION
**Pull request by stlaz**
Since the oauth-server's serving certificates can now be customized, it's important to publish the serving certificate of the server. This certificate should be used by applications like the oauth-proxy and openshift-console in order for them to be capable of successfully complete the OAuth2 authorization process as these clients eventually get to talk directly to the oauth-server and therefore need to trust it.

/assign @slaskawi 
/cc @deads2k @jhadvig @spadgett

### CODE REVIEW

File: manifests/04_oauthserver_trust_role.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  annotations:
+    rbac.authorization.kubernetes.io/autoupdate: "true"
+  name: system:openshift:oauth-servercert-trust
```
Comment (slaskawi) [inline manifests/04_oauthserver_trust_role.yaml:6]: I see the `system:openshift:operator:authentication` already binds to `cluster-admin`. So do we need this spacial file here? We already should have necessary permissions.

**Review by slaskawi — COMMENTED** [review]
LGTM in general. Just one thing to clarify with the regards to RBAC.

File: manifests/04_oauthserver_trust_role.yaml (hunk 1 above)
Comment (stlaz) [reply manifests/04_oauthserver_trust_role.yaml:6]: this role is not for the authentication operator, but for everyone (see the rolebinding below)

Comment (slaskawi) [reply manifests/04_oauthserver_trust_role.yaml:6]: For me it's fine. But please let someone more experienced have a look at this piece as well.

**Review by slaskawi — APPROVED** [review]
/lgtm

**Comment by stlaz** [comment]
/hold
for further review

**Comment by stlaz** [comment]
Added a more verbose error in the tests to indicate what's happening.

File: manifests/04_oauthserver_trust_role.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  annotations:
+    rbac.authorization.kubernetes.io/autoupdate: "true"
```
Comment (deads2k) [inline manifests/04_oauthserver_trust_role.yaml:5]: This should not be needed. We should not honor a "false" here, so it doesn't matter.

File: manifests/04_oauthserver_trust_role.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
```
Comment (deads2k) [inline manifests/04_oauthserver_trust_role.yaml:3]: missing namespace?  The namespace here is known.

File: manifests/04_oauthserver_trust_role.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (deads2k) [inline manifests/04_oauthserver_trust_role.yaml:1]: location on this file is incorrect.  This is about access to something managed *by* the operator, not something managed *for* the operator.  This should be created by the operator, not by the CVO.  Please move to bindata and use the staticresourcecontroller to create and maintain this.

File: pkg/controllers/common/ingress_config.go [hunk 5]
```diff
@@ -71,24 +71,29 @@ func GetCustomRouteHostname(ingress *configv1.Ingress, namespace string, name st
 // GetActiveRouterCertKeyBytes returns a byte array containing the server certificates, a byte array containing the private key,
 // a boolean representing if the default openshift-authentication/v4-0-config-system-router-certs secret is being used, and
 // any errors retrieving the active router secret.
-func GetActiveRouterCertKeyBytes(secretLister corev1listers.SecretLister, ingressConfig *configv1.Ingress, namespace string, defaultSecretName string, customSecretName string) ([]byte, []byte, bool, error) {
+func GetActiveRouterCertKeyBytes(
+	secretLister corev1listers.SecretLister,
+	ingressConfig *configv1.Ingress,
+	namespace string,
+	defaultSecretName string, customSecretName string,
+) (cert []byte, key []byte, isDefaultSecret bool, err error) {
```
Comment (deads2k) [inline pkg/controllers/common/ingress_config.go:79]: no named return values.

File: pkg/controllers/trustdistribution/trustdistribution_controller.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	var serverCert *x509.Certificate
+	var errs []error
+	for block, rest := pem.Decode(certBundle); block != nil; block, rest = pem.Decode(rest) {
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		if !cert.IsCA {
+			// we found our server cert, get outta here
```
Comment (deads2k) [inline pkg/controllers/trustdistribution/trustdistribution_controller.go:79]: seems like the entire chain of serving cert to root is valid for inclusion.

File: manifests/04_oauthserver_trust_role.yaml (hunk 3 above)
Comment (stlaz) [reply manifests/04_oauthserver_trust_role.yaml:3]: well spotted

File: bindata/oauth-openshift/trust_distribution_rolebinding.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: system:openshift:oauth-servercert-trust
+  namespace: openshift-config-managed
+  annotations:
```
Comment (deads2k) [inline bindata/oauth-openshift/trust_distribution_rolebinding.yaml:6]: unnecessary now

**Comment by stlaz** [comment]
/hold cancel
reapplying lgtm

**Pull request merged**